<a href="https://colab.research.google.com/github/hyun-sik-cho/LearningfromAudio/blob/master/LG_code.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 영업 성공 여부 분류 경진대회

## 1. 데이터 확인

### 필수 라이브러리

In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics import (
    accuracy_score,
    confusion_matrix,
    f1_score,
    precision_score,
    recall_score,
)
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier

### 데이터 셋 읽어오기

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
df_train = pd.read_csv("/content/drive/MyDrive/LG_data/train.csv") # 학습용 데이터
df_test = pd.read_csv("/content/drive/MyDrive/LG_data/submission.csv") # 테스트 데이터(제출파일의 데이터)

In [4]:
df_train.head() # 학습용 데이터 살펴보기

bant_submit          customer_country business_unit  com_reg_ver_win_rate  \
0          1.0  /Quezon City/Philippines            AS              0.066667   
1          1.0        /PH-00/Philippines            AS              0.066667   
2          1.0           /Kolkata /India            AS              0.088889   
3          1.0        /Bhubaneswar/India            AS              0.088889   
4          1.0          /Hyderabad/India            AS              0.088889   

   customer_idx          customer_type  enterprise  historical_existing_cnt  \
0         32160           End-Customer  Enterprise                      NaN   
1         23122           End-Customer  Enterprise                     12.0   
2          1755           End-Customer  Enterprise                    144.0   
3          4919           End-Customer  Enterprise                      NaN   
4         17126  Specifier/ Influencer  Enterprise                      NaN   

   id_strategic_ver  it_strategic_ver  ...  response_corporate  \
0               NaN               NaN  ...               LGEPH   
1               NaN               NaN  ...               LGEPH   
2               NaN               NaN  ...               LGEIL   
3               NaN               NaN  ...               LGEIL   
4               NaN               NaN  ...               LGEIL   

    expected_timeline  ver_cus ver_pro ver_win_rate_x ver_win_ratio_per_bu  \
0  less than 3 months        1       0       0.003079             0.026846   
1  less than 3 months        1       0       0.003079             0.026846   
2  less than 3 months        1       0       0.003079             0.026846   
3  less than 3 months        1       0       0.003079             0.026846   
4  less than 3 months        0       0       0.003079             0.026846   

        business_area business_subarea lead_owner is_converted  
0  corporate / office      Engineering          0         True  
1  corporate / office      Advertising          1         True  
2  corporate / office     Construction          2         True  
3  corporate / office      IT/Software          3         True  
4  corporate / office              NaN          4         True  

[5 rows x 29 columns]

## 2. 데이터 전처리

### 레이블 인코딩

In [5]:
def label_encoding(series: pd.Series) -> pd.Series:
    """범주형 데이터를 시리즈 형태로 받아 숫자형 데이터로 변환합니다."""

    my_dict = {}

    # 모든 요소를 문자열로 변환
    series = series.astype(str)

    for idx, value in enumerate(sorted(series.unique())):
        my_dict[value] = idx
    series = series.map(my_dict)

    return series

In [6]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59299 entries, 0 to 59298
Data columns (total 29 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   bant_submit              59299 non-null  float64
 1   customer_country         58317 non-null  object 
 2   business_unit            59299 non-null  object 
 3   com_reg_ver_win_rate     14568 non-null  float64
 4   customer_idx             59299 non-null  int64  
 5   customer_type            15338 non-null  object 
 6   enterprise               59299 non-null  object 
 7   historical_existing_cnt  13756 non-null  float64
 8   id_strategic_ver         3444 non-null   float64
 9   it_strategic_ver         1121 non-null   float64
 10  idit_strategic_ver       4565 non-null   float64
 11  customer_job             40566 non-null  object 
 12  lead_desc_length         59299 non-null  int64  
 13  inquiry_type             58358 non-null  object 
 14  product_category      

In [7]:
# 레이블 인코딩할 칼럼들
label_columns = [
    "customer_country",
    "business_subarea",
    "business_area",
    "business_unit",
    "customer_type",
    "enterprise",
    "customer_job",
    "inquiry_type",
    "product_category",
    "product_subcategory",
    "product_modelname",
    "customer_country.1",
    "customer_position",
    "response_corporate",
    "expected_timeline",
]

df_all = pd.concat([df_train[label_columns], df_test[label_columns]])

for col in label_columns:
    df_all[col] = label_encoding(df_all[col])

다시 학습 데이터와 제출 데이터를 분리합니다.

In [8]:
for col in label_columns:
    df_train[col] = df_all.iloc[: len(df_train)][col]
    df_test[col] = df_all.iloc[len(df_train) :][col]

### 2-2. 학습, 검증 데이터 분리

In [9]:
x_train, x_val, y_train, y_val = train_test_split(
    df_train.drop("is_converted", axis=1),
    df_train["is_converted"],
    test_size=0.2,
    shuffle=True,
    random_state=400,
)

## 3. 모델 학습

### 모델 정의

In [10]:
model =xgboost()

### 모델 학습

In [11]:
model.fit(x_train.fillna(0), y_train)

DecisionTreeClassifier()

### 모델 성능 보기

In [12]:
def get_clf_eval(y_test, y_pred=None):
    confusion = confusion_matrix(y_test, y_pred, labels=[True, False])
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, labels=[True, False])
    recall = recall_score(y_test, y_pred)
    F1 = f1_score(y_test, y_pred, labels=[True, False])

    print("오차행렬:\n", confusion)
    print("\n정확도: {:.4f}".format(accuracy))
    print("정밀도: {:.4f}".format(precision))
    print("재현율: {:.4f}".format(recall))
    print("F1: {:.4f}".format(F1))

In [13]:
pred = model.predict(x_val.fillna(0))
get_clf_eval(y_val, pred)

오차행렬:
 [[  737   210]
 [  261 10652]]

정확도: 0.9603
정밀도: 0.7385
재현율: 0.7782
F1: 0.7578


base line :

오차행렬:
 [[  737   210]
 [  261 10652]]

정확도: 0.9603
정밀도: 0.7385
재현율: 0.7782
F1: 0.7578

## 4. 제출하기

### 테스트 데이터 예측

In [14]:
# 예측에 필요한 데이터 분리
x_test = df_test.drop(["is_converted", "id"], axis=1)

In [15]:
test_pred = model.predict(x_test.fillna(0))
sum(test_pred) # True로 예측된 개수

1165

### 제출 파일 작성

In [17]:
# 제출 데이터 읽어오기 (df_test는 전처리된 데이터가 저장됨)
df_sub = pd.read_csv("/content/drive/MyDrive/LG_data/submission.csv")
df_sub["is_converted"] = test_pred

# 제출 파일 저장
df_sub.to_csv("submission.csv", index=False)

**우측 상단의 제출 버튼을 클릭해 결과를 확인하세요**